File: Problems/0322.Coin-Change/CoinChange.py
from typing import List
from functools import lru_cache


class Solution:

    # Runtime: 1444 ms, faster than 55.25% of Python3 online submissions for Coin Change.
    #  O(kn)
    def coinChange(self, coins: List[int], amount: int) -> int:
        @lru_cache(maxsize=None)
        def dp(n):
            if n == 0: return 0
            if n < 0:  return -1
            res = float('inf')
            for coin in coins:
                sub_cnt = dp(n - coin)
                if sub_cnt == -1:
                    continue
                res = min(res, 1 + sub_cnt)
            return res == float('inf') and -1 or res

        return dp(amount)

    # Runtime: 1340 ms, faster than 67.25% of Python3 online submissions for Coin Change.
    def coinChangeV2(self, coins: List[int], amount: int) -> int:
        dp = [0] + [amount+1] * amount
        for i in range(1, amount+1):
            for coin in coins:
                if i - coin < 0: continue
                dp[i] = min(dp[i], 1 + dp[i - coin])
        return dp[amount] == amount+1 and -1 or dp[amount]


    # Runtime: 820 ms, faster than 95.46% of Python3 online submissions for Coin Change.
    def coinChangeV3(self, coins: List[int], amount: int) -> int:
        dp = [amount+1] * (amount +1)
        dp[0] = 0
        for i in range(1, amount+1):
            for coin in coins:
                if i - coin < 0: continue
                sub_res = 1 + dp[i - coin]
                if sub_res < dp[i]:
                    dp[i] = sub_res
        return dp[amount] == amount+1 and -1 or dp[amount]


if __name__ == "__main__":
    print(Solution().coinChangeV2([1, 2, 5], 11))

File: leetcode/0322.Coin-Change/CoinChange.go
package leetcode

// Runtime: 8 ms, faster than 90.05% of Go online submissions for Coin Change.
func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1, amount+1)
	for i := range dp {
		dp[i] = amount + 1
	}
	dp[0] = 0
	for i := 1; i <= amount; i++ {
		for _, coin := range coins {
			if i-coin < 0 {
				continue
			}
			if subRes := 1 + dp[i-coin]; subRes < dp[i] {
				dp[i] = subRes
			}
		}
	}
	if dp[amount] == amount+1 {
		return -1
	}
	return dp[amount]
}

// Runtime: 4 ms, faster than 99.73% of Go online submissions for Coin Change
func coinChangeV2(coins []int, amount int) int {
	dp := make([]int, amount+1, amount+1)
	for i := range dp {
		dp[i] = amount + 1
	}
	dp[0] = 0

	for i := 1; i <= amount; i++ {
		for j := range coins {
			coin := coins[j]
			if i-coin < 0 {
				continue
			}
			if subRes := 1 + dp[i-coin]; subRes < dp[i] {
				dp[i] = subRes
			}
		}
	}
	if dp[amount] == amount+1 {
		return -1
	}
	return dp[amount]
}
